README
##############################################################
# 
# CentOS 8 LedgerSMB Easy Installer
# 
# By: STEVE ARBOUR 	
# Special thanks go to: EHUELSMANN, HASORLI, DCG, YVES 
# License: GNU GENERAL PUBLIC LICENSE, Version 2, June 1991
# 
##############################################################
# 
# STEP REQUIRED: 
#
# - INSTALL CENTOS 8 - SERVER NO GUI
# - UNPACK THIS INSTALLER
# - READ THIS FILE FOR USAGE SYNTAX.
# - EDIT THE FILE CONFIGURATION
# - MAKE THE INSTALLER EXECUTABLE (chmod ug+x install.sh)
# - MAKE USE OF THE INSTALLER (install.sh)
# 
# 
# THIS SCRIPT REQUIRE:
#
# - AN EXTERNAL IP ADDRESS
# - A HOSTNAME THAT RESOLVE TO THAT IP ADDRESS
# - AN EMAIL ADDRESS FOR LETSENCRYPT TO USE.
#
# 
# For usage help and syntax, simply use './install.sh --help'
# 
# *** SECURITY WARNING ***: 
# Once the installer finish and you have tested the 
# installation, be sure to remove the CONFIGURATION file
# because it contain your LSMB_DBADMIN password used for 
# initial installation
# 
##############################################################
#
# COMMAND LINE ARGUMENTS:
#
# --selinux-mode=  enforcing, permissive, disabled
# --skip-selinux=  yes, no, 1, 0
# --ipv6=  enabled, disabled
# --branch=  1.5, 1.6, 1.7, stable, master
# --update-all=  yes, no, 1, 0
# --reboot-after-install=  yes, no, 1, 0
# --skip-httpd-install=  yes, no, 1, 0
# --skip-httpd-config=  yes, no, 1, 0
# --httpd-enable-http=  yes, no, 1, 0
# --httpd-enable-https=  yes, no, 1, 0
# --skip-letsencrypt-install=  yes, no, 1, 0
# --skip-letsencrypt-config=  yes, no, 1, 0
# --skip-postgresql-install=  yes, no, 1, 0
# --skip-postgresql-config=  yes, no, 1, 0
# --skip-firewall-port80=  yes, no, 1, 0
# --skip-firewall-port443=  yes, no, 1, 0
# --skip-cpan-config=  yes, no, 1, 0
#
##############################################################

CHANGELOG
Version 1.1.1
-	Fixed a bug with the arguments processing.

Version 1.1.0
- 	Major overhaul of the script code, optimization, better integration.
- 	Added new set of command line arguments to allow a fully customizable 
	and perhaps scriptable installation
-	Updated the README and the help section
- 	Improved compatibility when installing 1.5/1.6

New branch 1.1.x


Version 1.0.5
-   Enable the CentOS Powertools Repository to access missing packages for an 
    easy and faster installation.
-   Added more package installation via DNF from various repos, including the 
    newly added CentOS Powertools.
-   Removed the Building of Package via CPAN since it isn't necessary anymore
    that allow to save lot of time during installation.
-   Updated the README and Last message of the installation to suggest removing
    the CONFIGURATION file for security reason.

Version 1.0.4
-   Added auto-detection of the lastest stable version on a specific branch.
-   Added Color to the install.sh script.
-   Added Color to the help section.
-   Updated the README file.
-   Added Experimental support for LetEncrypts Dry-Mode in the CONFIGURATION file
    (This allow to test-run before actually submitting real certificates request)
    $LEDGERSMB_LETENCRYPTS_DRYRUN.
-   Redefining STABLE as the lastest stable of the lastest branch.
-   Added credits to ehuelsmann hasorli dcg Yves.
-   General Script optimization.
-   SELinux fix when enforcing.

Version 1.0.3
-   Added support for the STABLE branch on git.
-   Made the STABLE branch a default.

Version 1.0.2
-   Removed the section of the script that used rc.local since we only 
    do one reboot, this is not necessary.
-   Removed an unused line when setting selinux to permissive or disabled.
-   Added an command line argument for the ledgersmb branch version and 
    was tested with 1.7 and master, which was 1.8 at the time of the testing.
-   Removed the multiple script files and made one larger single install.sh file.
-   Created a display_help function to display help and manage the help message
    in one single place.
-   BUGFIX: Help message was not being displayed when wrong/invalid arguments was
    pass to the installer. 

Version 1.0.1
-   Removed the multiples reboots and only keep one reboot at the end.
-   Added parameters to the command line installer.
-   Added usage help to the command line installer.
-   Updated the README file.
-   Included SELINUX configuration for ledgersmb.
-   Temporary disable selinux until reboot if needed.
-   Temporary disable ipv6 until reboot if needed.
-   Created a CHANGELOG file.

Version 1.0.0
-   First beta version.

step3.sh
#!/bin/bash

WORKING_INSTALLATION_PATH="`dirname \"$0\"`"
WORKING_INSTALLATION_PATH="`( cd \"$WORKING_INSTALLATION_PATH\" && pwd )`"
. $WORKING_INSTALLATION_PATH/CONFIGURATION

clear
echo " "
echo "-> INSTALLER STEP 3 INITIATED : PACKAGES AND NET/DB/LEDGERSMB SERVICE INSTALLATION PART 1"
echo " "
sleep 3


# FIRST ROUND OF PACKAGES INSTALLATION
dnf -y install nano gcc make wget git net-tools cpan cpanminus perl epel-release
dnf -y install httpd 
dnf -y install postgresql postgresql-devel postgresql-server
dnf -y install perl-CGI-Emulate-PSGI perl-CGI-Simple perl-Config-IniFiles
dnf -y install perl-DBD-Pg perl-DBI perl-DateTime perl-DateTime-Format-Strptime
dnf -y install perl-Digest-MD5 perl-File-MimeInfo perl-JSON-XS
dnf -y install perl-Locale-Maketext perl-Locale-Maketext-Lexicon
dnf -y install perl-Log-Log4perl perl-MIME-Base64 perl-MIME-Lite perl-Math-BigInt-GMP
dnf -y install perl-Moose perl-Number-Format perl-Plack perl-Template-Toolkit
dnf -y install perl-namespace-autoclean perl-MooseX-NonMoose perl-XML-Simple
dnf -y install perl-TeX-Encode texlive
dnf -y install libpqxx libpqxx-devel
dnf -y install perl-JSON-MaybeXS
dnf -y install perl-Starman
dnf -y install perl-CGI-Emulate-PSGI perl-CGI-Simple perl-Config-IniFiles perl-DBD-Pg perl-DBI perl-DateTime perl-DateTime-Format-Strptime perl-Digest-MD5 perl-File-MimeInfo perl-JSON-XS perl-Locale-Maketext perl-Locale-Maketext-Lexicon perl-Log-Log4perl perl-MIME-Base64 perl-MIME-Lite perl-Math-BigInt-GMP perl-Moose perl-Number-Format perl-Plack perl-Template-Toolkit perl-namespace-autoclean perl-MooseX-NonMoose perl-XML-Simple
dnf -y install expat-devel
dnf -y install texlive-latex
dnf -y install redhat-lsb
dnf -y install nodejs nodejs-devel nodejs-packaging nodejs-docs
dnf -y install java-latest-openjdk

# CONFIGURING FIREWALL
firewall-cmd --zone=public --add-port=80/tcp --permanent
firewall-cmd --zone=public --add-port=443/tcp --permanent
systemctl restart firewalld

# ACTIVATE HTTPD
systemctl enable httpd
systemctl start httpd

# POSTGRESQL
systemctl enable postgresql
postgresql-setup --initdb --unit postgresql
systemctl start postgresql
eval "su - postgres -c 'yes "$LEDGERSMB_LSMB_DBADMIN_PASS" | createuser -S -d -r -l -P lsmb_dbadmin'"

cat >/var/lib/pgsql/data/pg_hba.conf <<EOL
local   all             postgres                         peer
local   all             all                              peer
host    all             postgres         127.0.0.1/32     reject
host    all             postgres        ::1/128      reject
host    postgres,template0,template1   lsmb_dbadmin         127.0.0.1/32     md5
host    postgres,template0,template1   lsmb_dbadmin         ::1/128      md5
host    postgres,template0,template1   all          127.0.0.1/32     reject
host    postgres,template0,template1   all          ::1/128      reject
host    all             all             127.0.0.1/32     md5
host    all             all             ::1/128          md5
EOL

systemctl restart postgresql

# LEDGERSMB PART 1
cd /usr/local
git clone --recurse https://github.com/ledgersmb/LedgerSMB /usr/local/ledgersmb
chown -R apache:apache /usr/local/ledgersmb/UI
cp /usr/local/ledgersmb/doc/conf/ledgersmb.conf.default /usr/local/ledgersmb/ledgersmb.conf
cp /usr/local/ledgersmb/doc/conf/systemd/ledgersmb_starman.service /etc/systemd/system/


$WORKING_INSTALLATION_PATH/step4.sh

# END OF SCRIPT - STEP 3



step5.sh
#!/bin/bash

WORKING_INSTALLATION_PATH="`dirname \"$0\"`"
WORKING_INSTALLATION_PATH="`( cd \"$WORKING_INSTALLATION_PATH\" && pwd )`"
. $WORKING_INSTALLATION_PATH/CONFIGURATION

clear
echo " "
echo "-> INSTALLER STEP 5 INITIATED : CPAN PACKAGES AND LEDGERSMB PART 2"
echo " "
sleep 3

# CPAN INSTALLATION OF PACKAGES

mkdir ~/.cpan/
mkdir ~/.cpan/CPAN/
cat >/root/.cpan/CPAN/MyConfig.pm <<EOL
\$CPAN::Config = {
  'applypatch' => q[],
  'auto_commit' => q[0],
  'build_cache' => q[100],
  'build_dir' => q[/root/.cpan/build],
  'build_dir_reuse' => q[0],
  'build_requires_install_policy' => q[yes],
  'bzip2' => q[/usr/bin/bzip2],
  'cache_metadata' => q[1],
  'check_sigs' => q[0],
  'cleanup_after_install' => q[0],
  'colorize_output' => q[0],
  'commandnumber_in_prompt' => q[1],
  'connect_to_internet_ok' => q[1],
  'cpan_home' => q[/root/.cpan],
  'ftp_passive' => q[1],
  'ftp_proxy' => q[],
  'getcwd' => q[cwd],
  'gpg' => q[/usr/bin/gpg],
  'gzip' => q[/usr/bin/gzip],
  'halt_on_failure' => q[0],
  'histfile' => q[/root/.cpan/histfile],
  'histsize' => q[100],
  'http_proxy' => q[],
  'inactivity_timeout' => q[0],
  'index_expire' => q[1],
  'inhibit_startup_message' => q[0],
  'keep_source_where' => q[/root/.cpan/sources],
  'load_module_verbosity' => q[none],
  'make' => q[/usr/bin/make],
  'make_arg' => q[],
  'make_install_arg' => q[],
  'make_install_make_command' => q[/usr/bin/make],
  'makepl_arg' => q[],
  'mbuild_arg' => q[],
  'mbuild_install_arg' => q[],
  'mbuild_install_build_command' => q[./Build],
  'mbuildpl_arg' => q[],
  'no_proxy' => q[],
  'pager' => q[/usr/bin/less],
  'patch' => q[/usr/bin/patch],
  'perl5lib_verbosity' => q[none],
  'prefer_external_tar' => q[1],
  'prefer_installer' => q[MB],
  'prefs_dir' => q[/root/.cpan/prefs],
  'prerequisites_policy' => q[follow],
  'recommends_policy' => q[1],
  'scan_cache' => q[atstart],
  'shell' => q[/bin/bash],
  'show_unparsable_versions' => q[0],
  'show_upload_date' => q[0],
  'show_zero_versions' => q[0],
  'suggests_policy' => q[0],
  'tar' => q[/usr/bin/tar],
  'tar_verbosity' => q[none],
  'term_is_latin' => q[1],
  'term_ornaments' => q[1],
  'test_report' => q[0],
  'trust_test_report_history' => q[0],
  'unzip' => q[/usr/bin/unzip],
  'urllist' => [q[http://www.cpan.org/]],
  'use_prompt_default' => q[0],
  'use_sqlite' => q[0],
  'version_timeout' => q[15],
  'wget' => q[/usr/bin/wget],
  'yaml_load_code' => q[0],
  'yaml_module' => q[YAML],
};
1;
__END__
EOL

#### 0013 - CPAN - PART 2 - IN ORDER
cpan install App:Cpan
cpan install PGObject PGObject::Type::BigFloat Number::Format Config::IniFiles PGObject::Simple
cpan install MIME::Lite Moose HTTP::Exception PGObject::Simple::Role MooseX::NonMoose File::MimeInfo
cpan install PGObject::Type::ByteString Plack::Builder::Conditionals Plack::Middleware::Pod
cpan install Log::Log4perl Locale::Maketext::Lexicon DateTime::Format::Strptime PGObject::Type::DateTime
cpan install CGI::Emulate::PSGI CGI::Simple Digest::MD5 Encode File::Temp HTTP::Status List::Util Locale::Country Log::Log4Perl Mime::Base64 Try::Tiny Version::Compare
cpan install Net::Server XML::Parser XML::SAX::Expat XML::Simple
cpan install Plack::Middleware::Lint
cpan install Carp

$WORKING_INSTALLATION_PATH/step6.sh

# END OF SCRIPT - STEP 6



step6.sh
#!/bin/bash

WORKING_INSTALLATION_PATH="`dirname \"$0\"`"
WORKING_INSTALLATION_PATH="`( cd \"$WORKING_INSTALLATION_PATH\" && pwd )`"
. $WORKING_INSTALLATION_PATH/CONFIGURATION

clear
echo " "
echo "-> INSTALLER STEP 6 INITIATED : LEDGERSMB FINAL PART, UGLIFY-JS, DOJO"
echo " "
sleep 3

# LEDGERSMB PART 2
cd /usr/local/ledgersmb
sudo sed -i 's/^WorkingDirectory=.*/WorkingDirectory=\/usr\/local\/ledgersmb/g' /etc/systemd/system/ledgersmb_starman.service
sudo sed -i 's/^ExecStart=\/usr\/bin\/starman.*/ExecStart=\/usr\/local\/bin\/starman \\/g' /etc/systemd/system/ledgersmb_starman.service
useradd -d /non-existent -r -U -c "LedgerSMB/Starman service system user" ledgersmb

cd /usr/local/ledgersmb/UI/js-src/util/
npm install uglify-js@">=2.0 <3.0"

cd /usr/local/ledgersmb/
cpanm --quiet --notest --with-feature=starman --with-feature=latex-pdf-ps --with-feature=latex-pdf-images --installdeps /usr/local/ledgersmb/

cd /usr/local/ledgersmb/
make dojo

# END OF INSTALLATION
systemctl daemon-reload
systemctl enable postgresql
systemctl enable httpd
systemctl enable ledgersmb_starman
systemctl restart postgresql
systemctl restart httpd
systemctl restart ledgersmb_starman

# REMOVE ENVIRONMENT CONFIGURATION
. REMOVE_CONFIGURATION

clear
echo "Installation is (should be) complete"
echo " "
echo "Please open a compatible browser such as Mozilla Firefox, and point it to https://$LEDGERSMB_HOSTNAME/setup.pl to start using LedgerSMB"
echo " "
echo "This script was written by Steve Arbour"
echo " "
echo " "
echo "***************************************************************************************"
echo "***************************************************************************************"
echo "** WARNING THE SYSTEM WILL NOW REBOOT IN 15 SECONDS"
echo "***************************************************************************************"
echo "***************************************************************************************"
sleep 15
reboot

# END OF SCRIPT



step4.sh
#!/bin/bash

WORKING_INSTALLATION_PATH="`dirname \"$0\"`"
WORKING_INSTALLATION_PATH="`( cd \"$WORKING_INSTALLATION_PATH\" && pwd )`"
. $WORKING_INSTALLATION_PATH/CONFIGURATION

clear
echo " "
echo "-> INSTALLER STEP 4 INITIATED : HTTP, HTTPS AND LETENCRYPT"
echo " "
sleep 3

# HTTPD AND LETSENCRYPT
cat >/etc/httpd/conf.d/ledgersmb.conf <<EOL
<VirtualHost *:80>
  ServerName $LEDGERSMB_HOSTNAME
  ServerAlias $LEDGERSMB_HOSTNAME
  DocumentRoot /usr/local/ledgersmb/UI
  <Directory /usr/local/ledgersmb/UI>
      Options -Indexes +FollowSymLinks
      AllowOverride All
  </Directory>
  ErrorLog /var/log/httpd/ledgersmb-error.log
  CustomLog /var/log/httpd/ledgersmb-access.log combined
</VirtualHost>
EOL

systemctl restart httpd


#### 0010 - LET ENCRYPTS PART 1 : 
dnf -y install mod_ssl openssl yum-utils
dnf -y install python3-virtualenv python36-devel augeas-libs libffi-devel platform-python-devel python-rpm-macros python3-rpm-generators python3-wheel-wheel
openssl req -newkey rsa:4096 -x509 -sha256 -days 36500 -nodes -out /etc/pki/tls/certs/localhost.crt -keyout /etc/pki/tls/private/localhost.key -subj "$LEDGERSMB_SSL_PARAMETERS"
sleep 10
wget https://dl.eff.org/certbot-auto
sudo mv certbot-auto /usr/local/bin/certbot-auto
sudo chown root /usr/local/bin/certbot-auto
sudo chmod 0755 /usr/local/bin/certbot-auto
sudo /usr/local/bin/certbot-auto --apache -m $LEDGERSMB_LETENCRYPTS_EMAILADDR --agree-tos -n -d $LEDGERSMB_HOSTNAME
sleep 10
echo "0 0,12 * * * root python3 -c 'import random; import time; time.sleep(random.random() * 3600)' && /usr/local/bin/certbot-auto renew" | sudo tee -a /etc/crontab > /dev/null
systemctl restart httpd
sleep 10
#### 0011 - HTTPD PART 3: RESETUP HTTP AND HTTPS

cat >/etc/httpd/conf.d/ledgersmb.conf <<EOL
<VirtualHost *:80>
  ServerName $LEDGERSMB_HOSTNAME
  ServerAlias $LEDGERSMB_HOSTNAME
  DocumentRoot /usr/local/ledgersmb/UI
  <Directory /usr/local/ledgersmb/UI>
      Options -Indexes +FollowSymLinks
      AllowOverride All
  </Directory>
  ErrorLog /var/log/httpd/ledgersmb-error.log
  CustomLog /var/log/httpd/ledgersmb-access.log combined
</VirtualHost>
EOL

cat >/etc/httpd/conf.d/ledgersmb-le-ssl.conf <<EOL
<IfModule mod_ssl.c>
<VirtualHost *:443>
  ServerName $LEDGERSMB_HOSTNAME
  ServerAlias $LEDGERSMB_HOSTNAME
  DocumentRoot /usr/local/ledgersmb/UI
  <Directory /usr/local/ledgersmb/UI>
      Options -Indexes +FollowSymLinks
      AllowOverride All
  </Directory>
  Protocols h2 http:/1.1
  <If "%{HTTP_HOST} == '$LEDGERSMB_HOSTNAME_AUTOWWW'">
    Redirect permanent / https://$LEDGERSMB_HOSTNAME/
  </If>
  ErrorLog /var/log/httpd/$LEDGERSMB_HOSTNAME.ssl-error.log
  CustomLog /var/log/httpd/$LEDGERSMB_HOSTNAME-access.log combined
  SSLEngine On
  SSLCertificateFile /etc/letsencrypt/live/$LEDGERSMB_HOSTNAME/fullchain.pem
  SSLCertificateKeyFile /etc/letsencrypt/live/$LEDGERSMB_HOSTNAME/privkey.pem
  Include /etc/letsencrypt/options-ssl-apache.conf
  <Location "/">
    SSLRequireSSL
  </Location>
  Timeout 600
  ProxyTimeout 600
  RewriteEngine On
  RewriteRule "^/\$" "/login.pl" [R=301,L]
  RewriteRule "^/\." - [R=404,L]
  RewriteRule "\.conf\$" - [R=404,L]
  RewriteCond "%{REQUEST_FILENAME}" !-f
  RewriteCond "%{REQUEST_FILENAME}" !-d
  RewriteRule "^/(.*)" "http://127.0.0.1:5762/\$1" [P]
  ProxyPassReverse "/" "http://127.0.0.1:5762/"
</VirtualHost>
</IfModule>
EOL

systemctl restart httpd


$WORKING_INSTALLATION_PATH/step5.sh

# END OF SCRIPT - STEP 4

